rust-runtime/aws-smithy-client/src/timeout.rs

```diff
@@ -51,8 +62,163 @@ impl Settings {
+
+    fn layer(&self, inner: InnerService) -> Self::Service {
+        TimeoutService {
+            inner,
+            duration: self.0,
+        }
+    }
+}
+
+pin_project! {
+    #[non_exhaustive]
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    /// A future representing a timeout timer. Wraps a [Timeout] with extra context for error reporting
+    pub struct TimeoutLayerFuture<T> {
+        #[pin]
+        inner: Timeout<T, Sleep>
+    }
+}
+
+impl<InnerFuture, T, E> Future for TimeoutLayerFuture<InnerFuture>
+where
+    InnerFuture: Future<Output = Result<T, SdkError<E>>>,
+{
+    type Output = Result<T, SdkError<E>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let me = self.project();
+        match me.inner.poll(cx) {
+            Poll::Ready(Ok(t)) => Poll::Ready(t),
+            Poll::Ready(Err(timeout)) => {
+                Poll::Ready(Err(SdkError::ConstructionFailure(timeout.into())))
+            } // TODO: need to put this in the right place
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl<H, R, InnerService, E> tower::Service<Operation<H, R>> for TimeoutService<InnerService>
+where
+    InnerService: tower::Service<Operation<H, R>, Error = SdkError<E>>,
+{
+    type Response = InnerService::Response;
+    type Error = aws_smithy_http::result::SdkError<E>;
+    type Future = TimeoutLayerFuture<InnerService::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Operation<H, R>) -> Self::Future {
+        let base_future = self.inner.call(req);
+        let timeout_future = self.duration.map_or_else(
+            // If no timeout duration is provided, create a functionally infinite one
+            || default_async_sleep().expect("No default sleep impl was found. This is unexpected. Please report this bug.").sleep(Duration::MAX),
+            |duration| TokioSleep::new().sleep(duration),
+        );
+        let with_timeout = Timeout::new(base_future, timeout_future);
+        TimeoutLayerFuture {
+            inner: with_timeout,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::never::NeverService;
+    use crate::{SdkError, TimeoutLayer};
+    use aws_smithy_http::body::SdkBody;
+    use aws_smithy_http::operation::{Operation, Request};
+    use std::time::Duration;
+    use tower::{Service, ServiceBuilder, ServiceExt};
+
+    // Copied from aws-smithy-client/src/hyper_impls.rs
+    macro_rules! assert_elapsed {
+        ($start:expr, $dur:expr) => {{
+            let elapsed = $start.elapsed();
+            // type ascription improves compiler error when wrong type is passed
+            let lower: std::time::Duration = $dur;
+
+            // Handles ms rounding
+            assert!(
+                elapsed >= lower && elapsed <= lower + std::time::Duration::from_millis(5),
```

Comment:
we should upstream this, the provided macro in tokio test-util has too small of an epsilon